scarb/src/compiler/db.rs

```diff
@@ -59,6 +61,21 @@ fn load_plugins(
             let plugin = ws.config().cairo_plugins().fetch(package_id)?;
             let instance = plugin.instantiate()?;
             builder.with_plugin_suite(instance.plugin_suite());
+        } else if plugin_info.package.is_prebuilt() {
+            if proc_macros
+                .register_prebuilt(plugin_info.package.clone(), ws.config())
+                .is_err()
+            {
+                let plugin_unit = generate_cairo_plugin_compilation_units(&plugin_info.package)?
+                    .first()
+                    .unwrap()
+                    .clone();
+                ws.config()
+                    .ui()
+                    .print(Status::new("Compiling", &plugin_unit.name()));
+                compile_unit(plugin_unit, ws, false)?;
```

Comment:
In case of corrupted prebuilt file, this would cause the macro compilation to be run on each compilation unit that depends on the plugin. For instance in workspace of 5 packages, each with `lib` and `starknet-contract` targets, compilation of snforge plugin that failed to load as pre-built would be called 10 times (and print statuses to ui). 
Maybe instead we can return a typed error from `compile_unit_inner` function, e.g. `RetryWithPrerequisite ( Vec::<_>)` that would list additional compilation units that need to be run before this one? 
Then we can change `compile_units` from `for unit in units` to a FILO stack of CUs to compile, that in case of `RetryWithPrerequisite` would add the failed CU back to FILO with the returned prerequisites and try again.

Reply:
> Maybe instead we can return a typed error from `compile_unit_inner` function, e.g. `RetryWithPrerequisite ( Vec::<_>)` that would list additional compilation units that need to be run before this one?

`compile_unit_inner` does not load proc macros though.

I think the simplier approach would be to update `load_plugins` to just skip compilation if `shared_lib_path` already exists for the plugin. What do you think?

Reply:
Ok, you are right. 
But I still don't like calling `compile_unit` in `load_plugins` or any other place that is not `compile_units` 😛 It would make the execution flow a bit confusing, both from the code perspective and from the user one. For instance, we could end up printing "Compiling A", then print "Compiling B", then compile B and only then actually start compiling A. And what if we decided we need to compile some comp. units in parallel in future? 

> skip compilation if shared_lib_path already exists for the plugin

This is not necessarily true. For instance, when you depend on a local proc macro by path, it will always have the same target dir location in cache (name-path/version/), but it will require recompilation with Cargo on any changes made to it. 

I have another idea: 

Maybe we should add an optional field with type `Option<Arc<ProcMacroInstance>>` and name like `prebuilt` or smth to [`ProcMacroCompilationUnit`](https://github.com/software-mansion/scarb/blob/a219376ffd226bdc3ba1adaa9738cde206dc5551/scarb/src/compiler/compilation_unit.rs#L62) and [`CompilationUnitCairoPlugin`](https://github.com/software-mansion/scarb/blob/a219376ffd226bdc3ba1adaa9738cde206dc5551/scarb/src/compiler/compilation_unit.rs#L96)? We could then load prebuilts in [`generate_compilation_units`](https://github.com/software-mansion/scarb/blob/a219376ffd226bdc3ba1adaa9738cde206dc5551/scarb/src/ops/resolve.rs#L184C8-L184C34). If loading a prebuilt fails, we just set this field to `None`. The `compile_unit` could compile if the field is None, and just do nothing otherwise. Since it is an Arc, we would only load it once and then pass everywhere where needed. The `load_plugins` function can just use this field if present directly [as it just needs the Arc anyway](https://github.com/software-mansion/scarb/blob/f79011674512276dc164d354dce9017887c0cdb8/scarb/src/compiler/plugin/proc_macro/host.rs#L1155). This field would not be exposed in metadata, and we could add `skip_loading_prebuilts` to `generate_compilation_units` for the [metadata command](https://github.com/software-mansion/scarb/blob/f79011674512276dc164d354dce9017887c0cdb8/scarb/src/ops/metadata.rs#L47). 

WDYT about this?